packages/superset-ui-core/src/SupersetClient.js

```diff
@@ -0,0 +1,155 @@
+import callApi from './callApi';
+
+const AUTH_QUEUE_POLL_MS = 30;
+
+class SupersetClient {
+  constructor(config) {
+    const {
+      protocol = 'http',
+      host = '',
+      headers = {},
+      mode = 'same-origin',
+      timeout,
+      credentials,
+    } = config;
+
+    this.headers = headers;
+    this.host = host;
+    this.mode = mode;
+    this.timeout = timeout;
+    this.protocol = protocol;
+    this.credentials = credentials;
+    this.csrfToken = null;
+    this.didAuthSuccessfully = false;
+    this.csrfPromise = null;
+
+    // this.getUnauthorizedError = this.getUnauthorizedError.bind(this);
+    // this.waitForCSRF = this.waitForCSRF.bind(this);
+  }
+
+  isAuthenticated() {
+    return this.didAuthSuccessfully;
+  }
+
+  init() {
+    return this.getCSRFToken();
+  }
+
+  getCSRFToken() {
+    // If we can request this resource successfully, it means that the user has
+    // authenticated. If not we throw an error prompting to authenticate.
+    return (this.csrfPromise = callApi({
+      credentials: this.credentials,
+      headers: {
+        ...this.headers,
+      },
+      method: 'GET',
+      mode: this.mode,
+      timeout: this.timeout,
+      url: this.getUrl({ endpoint: 'superset/csrf_token/', host: this.host }),
+    }).then(
+      response => {
+        if (response.json) {
+          this.csrfToken = response.json.csrf_token;
+          this.headers = { ...this.headers, 'X-CSRFToken': this.csrfToken };
+          this.didAuthSuccessfully = !!this.csrfToken;
+        }
+
+        if (!this.csrfToken) {
+          return Promise.reject({ error: 'Failed to fetch CSRF token' });
+        }
+
+        return response;
```

Comment:
Should this return `response` or `this.csrfToken`?
Nobody is really using its result at this point, but seems like `this.csrfToken` will be more expected?

If somebody called this multiple times, it will send new request every time. Is that expected behavior? or if a promise already exists, then return `this.csrfPromise` so it will send request only once.